File: LLD/LLD-2/src/TicTakToeGame/src/main/java/org/example/strategies/winningStrategies/DiagonalWinningStrategy.java
package org.example.strategies.winningStrategies;

import org.example.models.Board;
import org.example.models.Move;
import org.example.models.Symbol;

import java.util.HashMap;
import java.util.Map;

public class DiagonalWinningStrategy implements WinningStrategy{
    private Map<Symbol, Integer> leftDiagonalMap = new HashMap<>();
    private Map<Symbol, Integer> rightDiagonalMap = new HashMap<>();

    @Override
    public boolean checkWinner(Move move, Board board) {
        int row = move.getCell().getRow();
        int col = move.getCell().getCol();

        Symbol symbol = move.getPlayer().getSymbol();
        if(row == col)
        {
            //cell is present on left diagonal
            if(leftDiagonalMap.containsKey(symbol))
            {
                leftDiagonalMap.put(symbol,leftDiagonalMap.get(symbol) + 1);
            }
            else
            {
                leftDiagonalMap.put(symbol, 1);
            }
        }
        if(row + col == board.getSize() -1)
        {
            if(rightDiagonalMap.containsKey(symbol))
            {
                rightDiagonalMap.put(symbol, rightDiagonalMap.get(symbol) + 1);
            }
            else
            {
                rightDiagonalMap.put(symbol, 1);
            }
        }

        if(row == col && leftDiagonalMap.get(symbol) == board.getSize())
            return true;
        if(row + col == board.getSize() - 1 && rightDiagonalMap.get(symbol) == board.getSize())
            return true;
        return false;
    }
}

File: LLD3/TicTacToe/TicTacToe/Strategies/WinningStrategies/DiagonalWinningStrategy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Threading.Tasks;
using TicTacToe.Models;

namespace TicTacToe.Strategies.WinningStrategies
{
    internal class DiagonalWinningStrategy : IWinningStrategy
    {
        public bool CheckForWinner(Move move, Board board)
        {

            int row = move.Cell.Row;
            int col = move.Cell.Col;
            bool left = false;
            bool right = false;
            if (row == col)
                left = LeftDiagonalCheck(move.Player.Symbol, board.Dimension);
            else if (row + col == board.Dimension)
                right = RightDiagonalCheck(move.Player.Symbol, board.Dimension);

            return left || right;
        }

        public bool LeftDiagonalCheck(Symbol symbol, int dimension)
        {
            Dictionary<Symbol, int> leftDiagonal = new Dictionary<Symbol, int>();

            if (!leftDiagonal.ContainsKey(symbol))
                leftDiagonal.Add(symbol, 1);
            else
                leftDiagonal[symbol]++;

            return leftDiagonal[symbol] == dimension;
        }

        public bool RightDiagonalCheck(Symbol symbol, int dimension)
        {

            Dictionary<Symbol, int> rightDiagonal = new Dictionary<Symbol, int>();

            if (!rightDiagonal.ContainsKey(symbol))
                rightDiagonal.Add(symbol, 1);
            else
                rightDiagonal[symbol]++;

            return rightDiagonal[symbol] == dimension;
        }
    }

}

